File: packages/go/oauth/device.go
package oauth

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// AuthorizeDevice initiates the device authorization flow.
// Following OAuth 2.0 Device Authorization Grant (RFC 8628).
func AuthorizeDevice(opts ...DeviceAuthOptions) (*DeviceAuthorizationResponse, error) {
	config := GetConfig()

	if config.ClientID == "" {
		return nil, fmt.Errorf("client ID is required for device authorization. Set OAUTH_CLIENT_ID or use Configure()")
	}

	// Build request body
	data := url.Values{}
	data.Set("client_id", config.ClientID)
	data.Set("scope", "openid profile email")

	// Add provider if specified (bypasses AuthKit login screen)
	if len(opts) > 0 && opts[0].Provider != "" {
		data.Set("provider", string(opts[0].Provider))
	}

	// Make request
	endpoint := GetDeviceAuthEndpoint()
	req, err := http.NewRequest("POST", endpoint, strings.NewReader(data.Encode()))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("device authorization request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		var errResp ErrorResponse
		if json.Unmarshal(body, &errResp) == nil && errResp.Error != "" {
			return nil, fmt.Errorf("device authorization failed: %s - %s", errResp.Error, errResp.ErrorDescription)
		}
		return nil, fmt.Errorf("device authorization failed: %s - %s", resp.Status, string(body))
	}

	var authResp DeviceAuthorizationResponse
	if err := json.Unmarshal(body, &authResp); err != nil {
		return nil, fmt.Errorf("failed to parse device authorization response: %w", err)
	}

	return &authResp, nil
}

// PollForTokens polls for tokens after device authorization.
// This implements the polling loop as specified in RFC 8628 Section 3.4.
//
// Parameters:
//   - deviceCode: Device code from authorization response
//   - interval: Polling interval in seconds (default: 5)
//   - expiresIn: Expiration time in seconds (default: 600)
func PollForTokens(deviceCode string, interval int, expiresIn int) (*TokenResponse, error) {
	config := GetConfig()

	if config.ClientID == "" {
		return nil, fmt.Errorf("client ID is required for token polling")
	}

	if interval <= 0 {
		interval = 5
	}
	if expiresIn <= 0 {
		expiresIn = 600
	}

	startTime := time.Now()
	timeout := time.Duration(expiresIn) * time.Second
	currentInterval := time.Duration(interval) * time.Second

	client := &http.Client{Timeout: 30 * time.Second}
	endpoint := GetTokenEndpoint()

	for {
		// Check if expired
		if time.Since(startTime) > timeout {
			return nil, fmt.Errorf("device authorization expired. Please try again")
		}

		// Wait for interval
		time.Sleep(currentInterval)

		// Build request
		data := url.Values{}
		data.Set("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
		data.Set("device_code", deviceCode)
		data.Set("client_id", config.ClientID)

		req, err := http.NewRequest("POST", endpoint, strings.NewReader(data.Encode()))
		if err != nil {
			continue // Retry on request creation error
		}
		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

		resp, err := client.Do(req)
		if err != nil {
			continue // Retry on network error
		}

		body, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			continue // Retry on read error
		}

		// Success case
		if resp.StatusCode == http.StatusOK {
			var tokenResp TokenResponse
			if err := json.Unmarshal(body, &tokenResp); err != nil {
				return nil, fmt.Errorf("failed to parse token response: %w", err)
			}
			return &tokenResp, nil
		}

		// Parse error response
		var errResp ErrorResponse
		if json.Unmarshal(body, &errResp) != nil {
			errResp.Error = "unknown"
		}

		switch TokenError(errResp.Error) {
		case TokenErrorAuthorizationPending:
			// User hasn't completed authorization yet - continue polling
			continue

		case TokenErrorSlowDown:
			// Increase interval by 5 seconds as per RFC 8628
			currentInterval += 5 * time.Second
			continue

		case TokenErrorAccessDenied:
			return nil, fmt.Errorf("access denied by user")

		case TokenErrorExpiredToken:
			return nil, fmt.Errorf("device code expired")

		default:
			return nil, fmt.Errorf("token polling failed: %s", errResp.Error)
		}
	}
}

// PollForTokensWithCallback polls for tokens with progress callbacks.
// This is useful for CLI applications that want to show polling status.
func PollForTokensWithCallback(
	deviceCode string,
	interval int,
	expiresIn int,
	onPoll func(attempt int),
) (*TokenResponse, error) {
	config := GetConfig()

	if config.ClientID == "" {
		return nil, fmt.Errorf("client ID is required for token polling")
	}

	if interval <= 0 {
		interval = 5
	}
	if expiresIn <= 0 {
		expiresIn = 600
	}

	startTime := time.Now()
	timeout := time.Duration(expiresIn) * time.Second
	currentInterval := time.Duration(interval) * time.Second

	client := &http.Client{Timeout: 30 * time.Second}
	endpoint := GetTokenEndpoint()
	attempt := 0

	for {
		// Check if expired
		if time.Since(startTime) > timeout {
			return nil, fmt.Errorf("device authorization expired. Please try again")
		}

		// Wait for interval
		time.Sleep(currentInterval)
		attempt++

		// Call progress callback
		if onPoll != nil {
			onPoll(attempt)
		}

		// Build request
		data := url.Values{}
		data.Set("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
		data.Set("device_code", deviceCode)
		data.Set("client_id", config.ClientID)

		req, err := http.NewRequest("POST", endpoint, strings.NewReader(data.Encode()))
		if err != nil {
			continue
		}
		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

		resp, err := client.Do(req)
		if err != nil {
			continue
		}

		body, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			continue
		}

		if resp.StatusCode == http.StatusOK {
			var tokenResp TokenResponse
			if err := json.Unmarshal(body, &tokenResp); err != nil {
				return nil, fmt.Errorf("failed to parse token response: %w", err)
			}
			return &tokenResp, nil
		}

		var errResp ErrorResponse
		if json.Unmarshal(body, &errResp) != nil {
			errResp.Error = "unknown"
		}

		switch TokenError(errResp.Error) {
		case TokenErrorAuthorizationPending:
			continue
		case TokenErrorSlowDown:
			currentInterval += 5 * time.Second
			continue
		case TokenErrorAccessDenied:
			return nil, fmt.Errorf("access denied by user")
		case TokenErrorExpiredToken:
			return nil, fmt.Errorf("device code expired")
		default:
			return nil, fmt.Errorf("token polling failed: %s", errResp.Error)
		}
	}
}

File: packages/rust/oauth/src/device.rs
//! Device authorization flow for OAuth.do
//!
//! Implements OAuth 2.0 Device Authorization Grant (RFC 8628)

use crate::config::get_config;
use crate::types::{
    DeviceAuthorizationResponse, OAuthProvider, TokenError, TokenErrorResponse, TokenResponse,
};
use anyhow::{anyhow, Result};
use std::time::{Duration, Instant};

/// Options for device authorization
#[derive(Debug, Default)]
pub struct DeviceAuthOptions {
    /// OAuth provider to use directly (bypasses AuthKit login screen)
    pub provider: Option<OAuthProvider>,
}

/// Initiate device authorization flow
///
/// Following OAuth 2.0 Device Authorization Grant (RFC 8628)
///
/// # Arguments
///
/// * `options` - Optional settings including provider for direct OAuth
///
/// # Returns
///
/// Device authorization response with codes and URIs
///
/// # Example
///
/// ```rust,no_run
/// use oauth_do::device::{authorize_device, DeviceAuthOptions};
///
/// #[tokio::main]
/// async fn main() -> anyhow::Result<()> {
///     let response = authorize_device(DeviceAuthOptions::default()).await?;
///     println!("Visit: {}", response.verification_uri);
///     println!("Enter code: {}", response.user_code);
///     Ok(())
/// }
/// ```
pub async fn authorize_device(options: DeviceAuthOptions) -> Result<DeviceAuthorizationResponse> {
    let config = get_config();

    if config.client_id.is_empty() {
        return Err(anyhow!(
            "Client ID is required for device authorization. Set OAUTH_CLIENT_ID or configure with client_id"
        ));
    }

    let url = "https://auth.apis.do/user_management/authorize/device";

    let mut params = vec![
        ("client_id", config.client_id.clone()),
        ("scope", "openid profile email".to_string()),
    ];

    // Add provider if specified (bypasses AuthKit login screen)
    if let Some(provider) = options.provider {
        params.push(("provider", provider.as_str().to_string()));
    }

    let client = reqwest::Client::new();
    let response = client
        .post(url)
        .header("Content-Type", "application/x-www-form-urlencoded")
        .form(&params)
        .send()
        .await?;

    if !response.status().is_success() {
        let status = response.status();
        let error_text = response.text().await.unwrap_or_default();
        return Err(anyhow!(
            "Device authorization failed: {} - {}",
            status,
            error_text
        ));
    }

    let data: DeviceAuthorizationResponse = response.json().await?;
    Ok(data)
}

/// Poll for tokens after device authorization
///
/// # Arguments
///
/// * `device_code` - Device code from authorization response
/// * `interval` - Polling interval in seconds (default: 5)
/// * `expires_in` - Expiration time in seconds (default: 600)
///
/// # Returns
///
/// Token response with access token and optionally user info
///
/// # Example
///
/// ```rust,no_run
/// use oauth_do::device::{authorize_device, poll_for_tokens, DeviceAuthOptions};
///
/// #[tokio::main]
/// async fn main() -> anyhow::Result<()> {
///     let auth_response = authorize_device(DeviceAuthOptions::default()).await?;
///     let tokens = poll_for_tokens(
///         &auth_response.device_code,
///         auth_response.interval,
///         auth_response.expires_in,
///     ).await?;
///     println!("Access token: {}", tokens.access_token);
///     Ok(())
/// }
/// ```
pub async fn poll_for_tokens(
    device_code: &str,
    interval: u64,
    expires_in: u64,
) -> Result<TokenResponse> {
    let config = get_config();

    if config.client_id.is_empty() {
        return Err(anyhow!("Client ID is required for token polling"));
    }

    let start_time = Instant::now();
    let timeout = Duration::from_secs(expires_in);
    let mut current_interval = Duration::from_secs(interval);

    let client = reqwest::Client::new();
    let url = "https://auth.apis.do/user_management/authenticate";

    loop {
        // Check if expired
        if start_time.elapsed() > timeout {
            return Err(anyhow!("Device authorization expired. Please try again."));
        }

        // Wait for interval
        tokio::time::sleep(current_interval).await;

        let params = [
            ("grant_type", "urn:ietf:params:oauth:grant-type:device_code"),
            ("device_code", device_code),
            ("client_id", &config.client_id),
        ];

        let response = client
            .post(url)
            .header("Content-Type", "application/x-www-form-urlencoded")
            .form(&params)
            .send()
            .await?;

        if response.status().is_success() {
            let data: TokenResponse = response.json().await?;
            return Ok(data);
        }

        // Handle error responses
        let error_response: TokenErrorResponse = response
            .json()
            .await
            .unwrap_or(TokenErrorResponse {
                error: Some("unknown".to_string()),
                error_description: None,
            });

        let error = TokenError::from_str(&error_response.error.unwrap_or_else(|| "unknown".to_string()));

        match error {
            TokenError::AuthorizationPending => {
                // Continue polling
                continue;
            }
            TokenError::SlowDown => {
                // Increase interval by 5 seconds
                current_interval += Duration::from_secs(5);
                continue;
            }
            TokenError::AccessDenied => {
                return Err(anyhow!("Access denied by user"));
            }
            TokenError::ExpiredToken => {
                return Err(anyhow!("Device code expired"));
            }
            TokenError::Unknown(msg) => {
                return Err(anyhow!("Token polling failed: {}", msg));
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_device_auth_options_default() {
        let options = DeviceAuthOptions::default();
        assert!(options.provider.is_none());
    }

    #[test]
    fn test_oauth_provider_as_str() {
        assert_eq!(OAuthProvider::GitHubOAuth.as_str(), "GitHubOAuth");
        assert_eq!(OAuthProvider::GoogleOAuth.as_str(), "GoogleOAuth");
        assert_eq!(OAuthProvider::MicrosoftOAuth.as_str(), "MicrosoftOAuth");
        assert_eq!(OAuthProvider::AppleOAuth.as_str(), "AppleOAuth");
    }
}
